File: apps/mobile/src/api/chatHub.ts
import {
    HubConnectionBuilder,
    HubConnection,
    HttpTransportType,
    LogLevel,
} from '@microsoft/signalr';
import * as SecureStore from 'expo-secure-store';
import { BASE_URL } from './client';

// ── Server → Client event payloads ─────────────────────────────────

export interface ReceiveMessagePayload {
    conversationId: string;
    encryptedPayload: string; // base64-encoded byte[]
    senderId: string;
    timestamp: string;
    serverMessageId: string;
    clientMessageId: string;
}

export interface MessageDeliveredPayload {
    conversationId: string;
    messageId: string;
    clientMessageId: string;
}

export interface MessageReadPayload {
    conversationId: string;
    messageId: string;
    readBy: string;
}

export interface TypingUpdatePayload {
    conversationId: string;
    userId: string;
    isTyping: boolean;
}

export interface PresenceChangedPayload {
    userId: string;
    status: 'Online' | 'Offline';
    lastSeen: string;
}

// ── Event map for type-safe subscriptions ───────────────────────────

export interface ChatHubEvents {
    ReceiveMessage: (payload: ReceiveMessagePayload) => void;
    MessageDelivered: (payload: MessageDeliveredPayload) => void;
    MessageRead: (payload: MessageReadPayload) => void;
    TypingUpdate: (payload: TypingUpdatePayload) => void;
    PresenceChanged: (payload: PresenceChangedPayload) => void;
    Error: (message: string) => void;
}

export type ChatHubEventName = keyof ChatHubEvents;

// ── Connection singleton ────────────────────────────────────────────

let connection: HubConnection | null = null;

export function getConnection(): HubConnection | null {
    return connection;
}

export function createConnection(): HubConnection {
    if (connection) return connection;

    connection = new HubConnectionBuilder()
        .withUrl(`${BASE_URL}/hubs/chat`, {
            transport: HttpTransportType.WebSockets,
            skipNegotiation: true,
            accessTokenFactory: async () => {
                const token = await SecureStore.getItemAsync('accessToken');
                return token ?? '';
            },
        })
        .withAutomaticReconnect([0, 1_000, 5_000, 10_000, 30_000])
        .configureLogging(LogLevel.Warning)
        .build();

    return connection;
}

export async function startConnection(): Promise<void> {
    const conn = connection ?? createConnection();
    await conn.start();
}

export async function stopConnection(): Promise<void> {
    if (!connection) return;
    await connection.stop();
    connection = null;
}

// ── Invoke wrappers ─────────────────────────────────────────────────

export function sendMessage(
    conversationId: string,
    encryptedPayload: string,
    clientMessageId: string,
    messageType: string = 'Text',
): Promise<void> {
    if (!connection) throw new Error('Hub not connected');
    return connection.invoke(
        'SendMessage',
        conversationId,
        encryptedPayload,
        clientMessageId,
        messageType,
    );
}

export function markMessageRead(
    conversationId: string,
    messageId: string,
): Promise<void> {
    if (!connection) throw new Error('Hub not connected');
    return connection.invoke('MessageRead', conversationId, messageId);
}

export function startTyping(conversationId: string): Promise<void> {
    if (!connection) throw new Error('Hub not connected');
    return connection.send('StartTyping', conversationId);
}

export function stopTyping(conversationId: string): Promise<void> {
    if (!connection) throw new Error('Hub not connected');
    return connection.send('StopTyping', conversationId);
}

export function setOnline(): Promise<void> {
    if (!connection) throw new Error('Hub not connected');
    return connection.send('SetOnline');
}

File: apps/server/src/BasecampSocial.Api/Hubs/ChatHub.cs
using BasecampSocial.Api.Data;
using BasecampSocial.Api.Data.Entities;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;

namespace BasecampSocial.Api.Hubs;

/// <summary>
/// SignalR hub for real-time messaging, presence, and typing indicators.
/// All message payloads are encrypted — the server relays opaque blobs.
/// </summary>
[Authorize]
public class ChatHub : Hub
{
    private readonly AppDbContext _db;
    private readonly ILogger<ChatHub> _logger;

    public ChatHub(AppDbContext db, ILogger<ChatHub> logger)
    {
        _db = db;
        _logger = logger;
    }

    /// <summary>Called when a client connects. Joins all conversation groups.</summary>
    public override async Task OnConnectedAsync()
    {
        var userId = GetUserId();

        // Add user to all their conversation groups for broadcasting
        var conversationIds = await _db.ConversationMembers
            .Where(cm => cm.UserId == userId)
            .Select(cm => cm.ConversationId.ToString())
            .ToListAsync();

        foreach (var conversationId in conversationIds)
            await Groups.AddToGroupAsync(Context.ConnectionId, conversationId);

        // Notify presence
        await Clients.Others.SendAsync("PresenceChanged", new
        {
            UserId = userId,
            Status = "Online",
            LastSeen = DateTimeOffset.UtcNow
        });

        _logger.LogInformation("User {UserId} connected to ChatHub", userId);
        await base.OnConnectedAsync();
    }

    /// <summary>Called when a client disconnects.</summary>
    public override async Task OnDisconnectedAsync(Exception? exception)
    {
        var userId = GetUserId();

        // Update last seen
        var user = await _db.Users.FindAsync(userId);
        if (user is not null)
        {
            user.LastSeenAt = DateTimeOffset.UtcNow;
            await _db.SaveChangesAsync();
        }

        // Notify presence
        await Clients.Others.SendAsync("PresenceChanged", new
        {
            UserId = userId,
            Status = "Offline",
            LastSeen = DateTimeOffset.UtcNow
        });

        _logger.LogInformation("User {UserId} disconnected from ChatHub", userId);
        await base.OnDisconnectedAsync(exception);
    }

    /// <summary>
    /// Client sends an encrypted message. Server persists and relays to conversation members.
    /// </summary>
    public async Task SendMessage(Guid conversationId, byte[] encryptedPayload, string clientMessageId, string messageType = "Text")
    {
        var userId = GetUserId();

        // Verify membership
        var isMember = await _db.ConversationMembers
            .AnyAsync(cm => cm.ConversationId == conversationId && cm.UserId == userId);

        if (!isMember)
        {
            await Clients.Caller.SendAsync("Error", "You are not a member of this conversation.");
            return;
        }

        // Persist the encrypted message
        var message = new Message
        {
            Id = Guid.NewGuid(),
            ConversationId = conversationId,
            SenderId = userId,
            EncryptedPayload = encryptedPayload,
            MessageType = Enum.Parse<MessageType>(messageType),
            CreatedAt = DateTimeOffset.UtcNow
        };

        _db.Messages.Add(message);
        await _db.SaveChangesAsync();

        // Relay to all members in the conversation group
        await Clients.Group(conversationId.ToString()).SendAsync("ReceiveMessage", new
        {
            ConversationId = conversationId,
            message.EncryptedPayload,
            SenderId = userId,
            Timestamp = message.CreatedAt,
            ServerMessageId = message.Id,
            ClientMessageId = clientMessageId
        });

        // Confirm delivery to the sender
        await Clients.Caller.SendAsync("MessageDelivered", new
        {
            ConversationId = conversationId,
            MessageId = message.Id,
            ClientMessageId = clientMessageId
        });

        _logger.LogDebug("Message {MessageId} sent in conversation {ConversationId}", message.Id, conversationId);
    }

    /// <summary>Client indicates they've read messages in a conversation.</summary>
    public async Task MessageRead(Guid conversationId, Guid messageId)
    {
        var userId = GetUserId();

        await Clients.OthersInGroup(conversationId.ToString()).SendAsync("MessageRead", new
        {
            ConversationId = conversationId,
            MessageId = messageId,
            ReadBy = userId
        });
    }

    /// <summary>Client starts typing in a conversation.</summary>
    public async Task StartTyping(Guid conversationId)
    {
        var userId = GetUserId();
        await Clients.OthersInGroup(conversationId.ToString()).SendAsync("TypingUpdate", new
        {
            ConversationId = conversationId,
            UserId = userId,
            IsTyping = true
        });
    }

    /// <summary>Client stops typing in a conversation.</summary>
    public async Task StopTyping(Guid conversationId)
    {
        var userId = GetUserId();
        await Clients.OthersInGroup(conversationId.ToString()).SendAsync("TypingUpdate", new
        {
            ConversationId = conversationId,
            UserId = userId,
            IsTyping = false
        });
    }

    /// <summary>Client signals they are online.</summary>
    public async Task SetOnline()
    {
        var userId = GetUserId();
        await Clients.Others.SendAsync("PresenceChanged", new
        {
            UserId = userId,
            Status = "Online",
            LastSeen = DateTimeOffset.UtcNow
        });
    }

    private Guid GetUserId()
    {
        var sub = Context.User?.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value
            ?? Context.User?.FindFirst("sub")?.Value
            ?? throw new HubException("User ID claim not found.");

        return Guid.Parse(sub);
    }
}
